cloud-controller-manager/do/kubernetes_lb_validator_test.go

```diff
@@ -0,0 +1,295 @@
+			expectedAllowed: false,
+			resp:            newFakeNotFoundResponse(),
+			err:             newFakeNotFoundErrorResponse(),
+			expectedMessage: "failed to validate lb create, could not get validation response: " + newFakeNotFoundErrorResponse().Error(),
+		},
+		{
+			name: "Allow Update happy path",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2", annotations{}), fakeService("old-service", annotations{annDOLoadBalancerID: "test"}))},
+			expectedAllowed: true,
+			expectedMessage: "valid update request",
+		},
+		{
+			name: "Deny Update invalid configuration",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2", annotations{}), fakeService("old-service", annotations{annDOLoadBalancerID: "test"}))},
+			expectedAllowed: false,
+			resp:            newFakeUnprocessableResponse(),
+			err:             newFakeUnprocessableErrorResponse(),
+			expectedMessage: "failed to validate lb update: " + newFakeUnprocessableErrorResponse().Error(),
+		},
+		{
+			name: "Deny Update validation error",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2", annotations{}), fakeService("old-service", annotations{annDOLoadBalancerID: "test"}))},
+			expectedAllowed: false,
+			resp:            newFakeNotFoundResponse(),
+			err:             newFakeNotFoundErrorResponse(),
+			expectedMessage: "failed to validate lb update, could not get validation response: " + newFakeNotFoundErrorResponse().Error(),
+		},
+	}
+
+	for _, test := range testcases {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			// setup client
+			gClient := godo.NewFromToken("")
+			gClient.LoadBalancers = &fakeLBService{
+				createFn: func(context.Context, *godo.LoadBalancerRequest) (*godo.LoadBalancer, *godo.Response, error) {
+					return &godo.LoadBalancer{ID: "2", Name: "two"}, test.resp, test.err
+				},
+				updateFn: func(context.Context, string, *godo.LoadBalancerRequest) (*godo.LoadBalancer, *godo.Response, error) {
+					return &godo.LoadBalancer{ID: "2", Name: "two"}, test.resp, test.err
+				},
+			}
+			gClient.Regions = &fakeRegionsService{
+				listFn: func(context.Context, *godo.ListOptions) ([]godo.Region, *godo.Response, error) {
+					return []godo.Region{{Name: "nyc3", Slug: "nyc3"}}, newFakeOKResponse(), nil
+				}}
+
+			decoder, err := admission.NewDecoder(scheme)
+			if err != nil {
+				t.Fatalf("failed to initialize decoder %s", err)
+			}
+
+			ll := zap.New().WithName("webhook-validation-server")
+			ctrlruntimelog.SetLogger(ll)
+
+			validator := &KubernetesLBServiceValidator{
+				Log:     ll,
+				decoder: decoder,
+				GClient: gClient,
+			}
+
+			res := validator.Handle(context.TODO(), test.req)
+			if res.Allowed != test.expectedAllowed {
+				t.Fatalf("got %v, want %v", res.Allowed, test.expectedAllowed)
+			}
+			if res.Result.Reason != "" && string(res.Result.Reason) != test.expectedMessage {
```

Comment:
I think the current logic misses a case: if neither the reason nor the message get populated (due to a bug) while there is an expected message, the test passes even though it should not.

With my suggestion elsewhere to assert on the error status code and use the `test.expectedErrorStatusCode` field to determine which field to compare against, the required adjustment should become more straight-forward. It could look something like this (untested, so please verify):

```golang
gotMessage := string(res.Result.Reason)
label := "reason"
if test.expectedErrorStatusCode != 0 {
  gotMessage = res.Result.Message
  label = "message"
}

if gotMessage != test.expectedMessage {
  t.Fatalf("got %s %q, want %q", label, gotMessage, test.expectedMessage)
}
```